### PR TITLE
add nested tensor tutorial

### PR DESCRIPTION
**Pull request by YifanShenSZ**
This is a beginner tutorial introducing the semantics of nested tensor. Operations are demonstrated through basic examples and a more realistic example (multi-head attention).

Thanks @cpuhrsch for pointing out torch.nn.MultiheadAttention, it is also included in the tutorial to demo how a fused black-box kernel can now be implemented transparently through basic operations.

No training is involved.

This is put under prototype due to nested-tensor ops have to wait for the next PyTorch stable release. Will move to beginner then.

### CODE REVIEW

**Comment by YifanShenSZ** [comment]
request review from @jbschlosser

File: beginner_source/nestedtensor_tutorial.py [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+
+Nested tensor is very similar to regular tensor, except for the shape:
+
+* for a regular tensor, each dimension has a size
+
+* for a nested tensor, not all dimensions have regular sizes: some of them are jagged
+
+Nested tensor arises from many application scenarios:
+
+* in NLP, sentenses have variable lengthes, so a batch of sentenses forms a nested tensor
```
Comment (cpuhrsch) [inline beginner_source/nestedtensor_tutorial.py:13]: nit: sentences

File: beginner_source/nestedtensor_tutorial.py [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+######################################################################
+# **others**
+#
+# Other operations have same semantics as regular tensors.
+nt_mm = torch.nested_tensor([torch.randn((2, 3, 4)), torch.randn((2, 3, 5))], device=device)
+nt3 = torch.matmul(nt2, nt_mm)
+print(nt3)
+
+######################################################################
+# Why Nested Tensor
```
Comment (cpuhrsch) [inline beginner_source/nestedtensor_tutorial.py:72]: nit: I've usually used `NestedTensor` for the name of the data structure, `nested_tensor` for the factory function and `nestedtensor` for the name of the project. See

**Comment by cpuhrsch** [comment]
This is great!

File: beginner_source/nestedtensor_tutorial.py (hunk 2 above)
Comment (YifanShenSZ) [reply beginner_source/nestedtensor_tutorial.py:72]: Is there a link to some reference?

Comment (cpuhrsch) [reply beginner_source/nestedtensor_tutorial.py:72]: I was going to add something, but now I can't find it anymore.

File: beginner_source/nestedtensor_tutorial.py [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+print("nested and padded calculations differ by", (out_nested.to_padded_tensor(0.0, (N, L_t, E_out)) - out_padded).abs().max().item())
+print("nested tensor multi-head attention takes", t1 - t0, "seconds")
+print("padded tensor multi-head attention takes", t2 - t1, "seconds")
+
+######################################################################
+# Nested tensor version has no redundancy,
+# so in sequential CPU execution it is faster than padded tensor version as expected.
+# Optimization for multi-threaded environment is underway.
+#
+# For now, performant kernels are provided for specific usages, e.g.
```
Comment (jbschlosser) [inline beginner_source/nestedtensor_tutorial.py:405]:
```suggestion
# For now, performant kernels are provided for specific use cases, e.g.
```

File: beginner_source/nestedtensor_tutorial.py [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+    b_q=b_q, b_k=b_k, b_v=b_v, b_out=b_out,
+    dropout_p=dropout_p)
+t2 = timeit.default_timer()
+
+print("nested and padded calculations differ by", (out_nested.to_padded_tensor(0.0, (N, L_t, E_out)) - out_padded).abs().max().item())
+print("nested tensor multi-head attention takes", t1 - t0, "seconds")
+print("padded tensor multi-head attention takes", t2 - t1, "seconds")
+
+######################################################################
+# Nested tensor version has no redundancy,
```
Comment (jbschlosser) [inline beginner_source/nestedtensor_tutorial.py:401]:
```suggestion
# The nested tensor version avoids wasted computation on padding,
```

File: beginner_source/nestedtensor_tutorial.py [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+
+######################################################################
+# Nested Tensor Initialization
+# ----------------
+#
+
+######################################################################
+# From python frontend, for now nested tensor can only be created from a list of tensors.
+nt = torch.nested_tensor([torch.randn((2, 6)), torch.randn((3, 6))], device=device)
+print(nt)
```
Comment (jbschlosser) [inline beginner_source/nestedtensor_tutorial.py:35]: Wdyt about demonstrating simple indexing here as well?

File: beginner_source/nestedtensor_tutorial.py [hunk 6]
```diff
@@ -0,0 +1,449 @@
...
+
+Nested tensor arises from many application scenarios:
+
+* in NLP, sentences have variable lengthes, so a batch of sentences forms a nested tensor
+
+* in CV, images have variable shapes, so a batch of images forms a nested tensor
+
+For practical reason, conceptually we implement nested tensor
+as a batch of tensors with different shapes,
+i.e. dimension 0 is assumed to be the batch dimension.
```
Comment (jbschlosser) [inline beginner_source/nestedtensor_tutorial.py:19]: This block of text seems a bit out-of-place here - maybe it's more useful under Nested Tensor Initialization when showing how to create an NT from a list of tensors?

To transition smoothly from the initial introduction to the meat of the tutorial, maybe a fluff line could be added here instead. Something like:
```
In this tutorial, we will demonstrate basic usage of nested tensors and motivate their usefulness
for operating on sequential data of varying lengths with a real-world example.
```

**Review by jbschlosser — COMMENTED** [review]
Thanks for all your hard work in creating both this tutorial and support for the ops mentioned! I added some minor nits related to formatting, grammar, and organization but the content is overall very solid!

**Comment by bhosmer** [comment]
@YifanShenSZ chiming in to say this is going to be an excellent resource, and a great way to unlock the value of your implementation work. Thanks for doing it!

File: beginner_source/nestedtensor_tutorial.py [hunk 7]
```diff
@@ -0,0 +1,449 @@
...
+#
+
+######################################################################
+# Operation coverage for nested tensor is narrower than regular tensors.
+# For now only basic operations such as index, dropout, softmax, transpose, reshape, linear, bmm are covered.
+# If you need certain operations, please file an `issue <https://github.com/pytorch/nestedtensor>`__.
+#
+# **reshape**
+#
+# Note that -1 no longer infers as the case of regular tensor.
```
[the remark this answers is not in the document]
Comment (YifanShenSZ) [reply beginner_source/nestedtensor_tutorial.py]: https://github.com/pytorch/tutorials/pull/1975/files#diff-60d5bdffd1a7c7991a41cd6db93fddf22dcfc7ee4bb29816053785b912d74f0aR66-R77

File: beginner_source/nestedtensor_tutorial.py [hunk 8]
```diff
@@ -0,0 +1,449 @@
...
+# Instead, it inherits the old size (here 2 for ``nt[0]`` and 3 for ``nt[1]``).
+# -1 is the only legal size to specify for a jagged dimension.
+nt1 = nt.reshape(2, -1, 2, 3)
+print(nt1)
+
+######################################################################
+# **transpose**
```
[the remark this answers is not in the document]
Comment (YifanShenSZ) [reply beginner_source/nestedtensor_tutorial.py:82]: https://github.com/pytorch/tutorials/pull/1975/files#diff-60d5bdffd1a7c7991a41cd6db93fddf22dcfc7ee4bb29816053785b912d74f0aR82-R89

File: beginner_source/nestedtensor_tutorial.py [hunk 9]
```diff
@@ -0,0 +1,449 @@
...
+print(nt2)
+
+######################################################################
+# **others**
+#
+# Other operations have same semantics as regular tensors.
+nt_mm = torch.nested_tensor([torch.randn((2, 3, 4)), torch.randn((2, 3, 5))], device=device)
+nt3 = torch.matmul(nt2, nt_mm)
+print(nt3)
+
```
[the remark this answers is not in the document]
Comment (YifanShenSZ) [reply beginner_source/nestedtensor_tutorial.py:109]: https://github.com/pytorch/tutorials/pull/1975/files#diff-60d5bdffd1a7c7991a41cd6db93fddf22dcfc7ee4bb29816053785b912d74f0aR36-R40

File: beginner_source/nestedtensor_tutorial.py [hunk 10]
```diff
@@ -0,0 +1,449 @@
...
+nt2 = nt1.transpose(1, 2)
+print(nt2)
+
+######################################################################
+# **others**
+#
+# Other operations have same semantics as regular tensors.
+nt_mm = torch.nested_tensor([torch.randn((2, 3, 4)), torch.randn((2, 3, 5))], device=device)
+nt3 = torch.matmul(nt2, nt_mm)
+print(nt3)
```
[the remark this answers is not in the document]
Comment (YifanShenSZ) [reply beginner_source/nestedtensor_tutorial.py]: https://github.com/pytorch/tutorials/pull/1975/files#diff-60d5bdffd1a7c7991a41cd6db93fddf22dcfc7ee4bb29816053785b912d74f0aR100-R108

File: beginner_source/nestedtensor_tutorial.py (hunk 5 above)
Comment (YifanShenSZ) [reply beginner_source/nestedtensor_tutorial.py:34]: https://github.com/pytorch/tutorials/pull/1975/files#diff-60d5bdffd1a7c7991a41cd6db93fddf22dcfc7ee4bb29816053785b912d74f0aR42-R48

File: prototype_source/nestedtensor.py [hunk 11]
```diff
@@ -0,0 +1,492 @@
...
+print(nt)
+
+######################################################################
+# By padding every underlying tensor to the same shape,
+# a nested tensor can be converted to a regular tensor.
+pt = nt.to_padded_tensor(0.0)
+print(pt)
+
+######################################################################
+# For practical reason, conceptually we implement nested tensor
```
Comment (jbschlosser) [inline prototype_source/nestedtensor.py:46]:
```suggestion
# For practical reasons, conceptually we implement nested tensor
```

File: prototype_source/nestedtensor.py [hunk 12]
```diff
@@ -0,0 +1,492 @@
...
+# As each operation must be explicitly implemented for nested tensors,
+# operation coverage for nested tensors is currently narrower than that of regular tensors.
+# For now, only basic operations such as index, dropout, softmax, transpose, reshape, linear, bmm are covered.
+# However, coverage is being expanded rapidly.
+# If you need certain operations, please file an `issue <https://github.com/pytorch/pytorch>`__
+# to help us prioritize coverage.
+#
+# **reshape**
+#
+# Reshape is to change the shape of a tensor.
```
Comment (jbschlosser) [inline prototype_source/nestedtensor.py:71]:
```suggestion
# The reshape op is for changing the shape of a tensor.
```

File: prototype_source/nestedtensor.py [hunk 13]
```diff
@@ -0,0 +1,492 @@
...
+# Its full semantics for regular tensors can be found
+# `here <https://pytorch.org/docs/stable/generated/torch.reshape.html>`__.
+# For regular tensors, when specifying the new shape,
+# a single dimension may be -1, in which case it is inferred
+# from the remaining dimensions and the number of elements.
+#
+# The semantics for nested tensors is similar, except that -1 no longer infers.
```
Comment (jbschlosser) [inline prototype_source/nestedtensor.py:78]:
```suggestion
# The semantics for nested tensors are similar, except that -1 no longer infers.
```

File: prototype_source/nestedtensor.py [hunk 14]
```diff
@@ -0,0 +1,492 @@
...
+# Instead, it inherits the old size (here 2 for ``nt[0]`` and 3 for ``nt[1]``).
+# -1 is the only legal size to specify for a jagged dimension.
+nt1 = nt.reshape(2, -1, 2, 3)
+print(nt1)
+
+######################################################################
+# **transpose**
+#
+# Transpose is to swap two dimensions of a tensor.
```
Comment (jbschlosser) [inline prototype_source/nestedtensor.py:87]:
```suggestion
# The transpose op is for swapping two dimensions of a tensor.
```

File: prototype_source/nestedtensor.py [hunk 15]
```diff
@@ -0,0 +1,492 @@
...
+# Its full semantics can be found
+# `here <https://pytorch.org/docs/stable/generated/torch.transpose.html>`__.
+# Note that nested tensor dimension 0 is special:
+# assumed to be the batch dimension,
+# so transposing nested tensor dimension 0 is forbidden.
```
Comment (jbschlosser) [inline prototype_source/nestedtensor.py:92]:
```suggestion
# so transposes involving nested tensor dimension 0 are forbidden.
```

File: prototype_source/nestedtensor.py [hunk 16]
```diff
@@ -0,0 +1,492 @@
...
+print(nt1)
+
+######################################################################
+# **transpose**
+#
+# Transpose is to swap two dimensions of a tensor.
+# Its full semantics can be found
+# `here <https://pytorch.org/docs/stable/generated/torch.transpose.html>`__.
+# Note that nested tensor dimension 0 is special:
+# assumed to be the batch dimension,
```
Comment (jbschlosser) [inline prototype_source/nestedtensor.py:91]:
```suggestion
# it is assumed to be the batch dimension,
```

File: prototype_source/nestedtensor.py [hunk 17]
```diff
@@ -0,0 +1,492 @@
...
+nt1 = nt.reshape(2, -1, 2, 3)
+print(nt1)
+
+######################################################################
+# **transpose**
+#
+# Transpose is to swap two dimensions of a tensor.
+# Its full semantics can be found
+# `here <https://pytorch.org/docs/stable/generated/torch.transpose.html>`__.
+# Note that nested tensor dimension 0 is special:
```
Comment (jbschlosser) [inline prototype_source/nestedtensor.py:90]:
```suggestion
# Note that nested tensor dimension 0 is special;
```

**Review by jbschlosser — COMMENTED** [review]
Couple more easy-to-fix nits then we're good!

**Review by jbschlosser — APPROVED** [review]
LGTM!

**Pull request merged**